ref/java/src/main/java/org/bech32/SegwitAddress.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.commons.lang3.ArrayUtils;
+
+public class SegwitAddress {
+
+    public static Pair<Byte, byte[]> decode(String hrp, String addr) throws Exception {
+
+        Pair<byte[], byte[]> p = Bech32.bech32Decode(addr);
+
+        byte[] hrpgot = p.getLeft();
+        String hrpgotStr =  new String(hrpgot);
+        if(hrpgot == null || hrpgotStr == null)  {
+          return Pair.of(null, null);
+        }
+        if (!hrp.equals(hrpgotStr))    {
+            throw new Exception("mismatching bech32 human readeable part");
+        }
+        if (!hrpgotStr.equalsIgnoreCase("bc") && !hrpgotStr.equalsIgnoreCase("tb"))    {
+            throw new Exception("invalid segwit human readable part");
+        }
+
+        byte[] data = p.getRight();
+        byte[] decoded = convertBits(Arrays.asList(ArrayUtils.toObject(Arrays.copyOfRange(data, 1, data.length))), 5, 8, false);
+        if(decoded.length < 2 || decoded.length > 40)   {
+            throw new Exception("invalid decoded data length");
+        }
+
+        byte witnessVersion = data[0];
+        if (witnessVersion > 16)   {
+            throw new Exception("invalid decoded witness version");
+        }
+
+        if (witnessVersion == 0 && decoded.length != 20 && decoded.length != 32)   {
+            throw new Exception("decoded witness version 0 with unknown length");
+        }
+
+        return Pair.of(witnessVersion, decoded);
+    }
+
+    public static String encode(byte[] hrp, byte witnessVersion, byte[] witnessProgram) throws Exception    {
+
+        byte[] prog = convertBits(Arrays.asList(ArrayUtils.toObject(witnessProgram)), 8, 5, true);
+        byte[] data = new byte[1 + prog.length];
+
+        System.arraycopy(new byte[] { witnessVersion }, 0, data, 0, 1);
+        System.arraycopy(prog, 0, data, 1, prog.length);
+
+        String ret = Bech32.bech32Encode(hrp, data);
+
+        return ret;
+    }
+
+    private static byte[] convertBits(List<Byte> data, int fromBits, int toBits, boolean pad) throws Exception    {
+
+        int acc = 0;
+        int bits = 0;
+        int maxv = (1 << toBits) - 1;
+        List<Byte> ret = new ArrayList<Byte>();
+
+        for(Byte value : data)  {
+
+            short b = (short)(value.byteValue() & 0xff);
+
+            if (b < 0) {
+                throw new Exception();
+            }
+            else if ((b >> fromBits) > 0) {
+                throw new Exception();
+            }
+            else    {
+                ;
+            }
+
+            acc = (acc << fromBits) | b;
+            bits += fromBits;
+            while (bits >= toBits)  {
+                bits -= toBits;
+                ret.add((byte)((acc >> bits) & maxv));
+            }
+        }
+
+        if(pad && (bits > 0))    {
+            ret.add((byte)((acc << (toBits - bits)) & maxv));
+        }
+        else if (bits >= fromBits || (byte)(((acc << (toBits - bits)) & maxv)) != 0)    {
+            throw new Exception("panic");
```

Comment:
This case is different from the cases above, as it can occur without any programmer error (malformed input can trigger it), so I'm unsure if an exception is the right way to deal with it.

Reply:
Updated.